File: Solutions/Python/ransom_note.py
"""
    [EASY]
    383. Ransom Note

    Concepts:
    - string
    - hash table

    Stats:
        Runtime | 15 ms     [Beats 76.44%]
        Memory  | 16.80 MB  [Beats 100%]
"""

class Solution:
    def canConstruct(self, ransomNote: str, magazine: str) -> bool:
        rCount = Counter(ransomNote)
        mCount = Counter(magazine)

        for val in rCount:
            if not (val in mCount and mCount[val] >= rCount[val]):
                return False

        return True

File: Solutions/Cpp/ransom_note.cpp
/*
    [EASY]
    383. Ransom Note

    Concepts:
    - string
    - hash table

    Stats:
        Runtime | 3 ms      [Beats 62.48%]
        Memory  | 11.90 MB  [Beats 29.56%]
*/

class Solution {
public:
    bool canConstruct(string ransom_note, string magazine) {
        std::unordered_map<char, int> r_count = {};
        std::unordered_map<char, int> m_count = {};

        for (const char c : ransom_note)
            r_count[c]++;
        for (const char c : magazine)
            m_count[c]++;

        for (const auto [k, v] : r_count) {
            if (!(m_count.contains(k) && m_count[k] >= v))
                return false;
        }

        return true;
    }
};